src/ophyd_async/epics/pmac/_pmacTrajectory.py

```diff
@@ -0,0 +1,128 @@
+            self.motors[motor.csAxis] = motor
+        self._fly_start: float
+        self.cs = cs
+        super().__init__(prefix, cs, name=name)
+
+    async def _ramp_up_velocity_pos(
+        self, velocity: float, motor: PmacCSMotor, end_velocity
+    ):
+        # Assuming ramping to or from 0
+        max_velocity_acceleration_time = await motor.acceleration_time.get_value()
+        max_velocity = await motor.max_velocity.get_value()
+        accl_time = max_velocity_acceleration_time * end_velocity / max_velocity
+        disp = 0.5 * (velocity + end_velocity) * accl_time
+        return [disp, accl_time]
+
+    @AsyncStatus.wrap
+    async def prepare(self, scanSpecStack):
+        # Which Axes are in use?
+
+        scanSize = len(scanSpecStack[0])
+        scanAxes = scanSpecStack[0].axes()
+
+        self.profile = {}
+        for axis in scanAxes:
+            self.profile[axis.lower()] = []
+            if axis != "DURATION":
+                self.profile[axis.lower() + "_velocity"] = []
+
+        # Calc Velocity
+
+        for axis in scanAxes:
+            for i in range(scanSize - 1):
+                if axis != "DURATION":
+                    self.profile[axis.lower() + "_velocity"].append(
+                        (
+                            scanSpecStack[0].midpoints[axis][i + 1]
+                            - scanSpecStack[0].midpoints[axis][i]
+                        )
+                        / (scanSpecStack[0].midpoints["DURATION"][i])
+                    )
+                    self.profile[axis.lower()].append(
+                        scanSpecStack[0].midpoints[axis][i]
+                    )
+                else:
+                    self.profile[axis.lower()].append(
+                        int(scanSpecStack[0].midpoints[axis][i] / TICK_S)
+                    )
+            if axis != "DURATION":
+                self.profile[axis.lower()].append(
+                    scanSpecStack[0].midpoints[axis][scanSize - 1]
+                )
+                self.profile[axis.lower() + "_velocity"].append(0)
+            else:
+                self.profile[axis.lower()].append(
+                    int(scanSpecStack[0].midpoints[axis][scanSize - 1] / TICK_S)
+                )
+
+        # Calculate Starting Position to allow ramp up to velocity
+        self.initial_pos = {}
+        run_up_time = 0
+        for axis in scanAxes:
+            if axis != "DURATION":
+                run_up_disp, run_up_time = await self._ramp_up_velocity_pos(
+                    0,
+                    self.motors[axis.lower()],
+                    self.profile[axis.lower() + "_velocity"][0],
+                )
+                self.initial_pos[axis] = self.profile[axis.lower()][0] - run_up_disp
+        self.profile["duration"][0] += run_up_time / TICK_S
+
+        # Send trajectory to brick
+        for axis in scanAxes:
+            if axis != "DURATION":
+                self.profile_cs_name.set(self.cs)
```

Comment:
Then if we add to Motor:
```python
    def __init__(...):
        ...
        self.input_link = epics_signal_r(str, prefix + ".INP")
```

Then we can do somewhere above:
```python
input_links = asyncio.gather(*[axis.input_link.get_value() for axis in scanAxes])
cs_ports = set()
cs_axes: Dict[Motor, int] = {}
for axis in scanAxes:
    input_link = await axis.input_link.get_value()
    # Split "@asyn(PORT,num)" into ["PORT", "num"]
    split = input_link.split("(")[1].rstrip(")").split(",")
    cs_ports.add(split[0].strip())
    cs_axes[axis] = int(split[1].strip())
assert len(cs_ports) == 1, "Motors in more than one CS"
```

Reply:
I believe this would only work for compound motors. An Axis with straight through mapping, (p45 Sample X for example) does not have a compound motor and its link is '@asyn(PORT, num)' where PORT is the brick and num refers to its axis number on the brick rather than a CS assignment.

However from the brick and axis number I can then use '{PMAC_PREFIX}:M{num}:CsPort_RBV' and '{PMAC_PREFIX}:M{num}:CsAxis_RBV' to get the port and axis. But this feels quite long-winded...

Reply:
> I believe this would only work for compound motors. 

Correct

> An Axis with straight through mapping, (p45 Sample X for example) does not have a compound motor and its link is '@ASYN(PORT, num)' where PORT is the brick and num refers to its axis number on the brick rather than a CS assignment.

I forgot P45 didn't have all compound axes... I guess we'll have to solve this now.

> However from the brick and axis number I can then use '{PMAC_PREFIX}:M{num}:CsPort_RBV' and '{PMAC_PREFIX}:M{num}:CsAxis_RBV' to get the port and axis. But this feels quite long-winded...

Yes, that seems right. However I think I did the wrong thing in Malcolm attaching them to the motor. We should attach them to the generic PMAC object instead, in the same way as the EDM screens have an "Axes" table with deferred move, CS port and CS Axis. 

I think we need to discuss this with the `DeviceVector` part below, but for now could you attach these somewhere to the PMAC object so you can use them in the CS code above?